csi: add support for enabling read affinity for rbd volumes. (#11631)
rook/rook: Documentation/Storage-Configuration/Ceph-CSI/ceph-csi-drivers.md

```

* PVCs created using the new storageclass will be encrypted.

## Enable Read affinity for RBD volumes

Ceph CSI supports mapping RBD volumes with krbd options to allow
serving reads from an OSD in proximity to the client, according to

> OSD locations defined in the CRUSH map. The format of the krbd option

Reviewer: Can we not tell them the krbd format? It's an implementation detail, they can find it in the link to the krbd doc for more details.
```suggestion
OSD locations defined in the CRUSH map and topology labels on nodes.
```

Ceph CSI supports mapping RBD volumes with krbd options to allow
serving reads from an OSD in proximity to the client, according to
OSD locations defined in the CRUSH map. The format of the krbd option
is: `"read_from_replica=localize,crush_location=type1:value1|type2:value2"`.

Refer to the [krbd-options](https://docs.ceph.com/en/latest/man/8/rbd/#kernel-rbd-krbd-options)
for more details.

Execute the following steps:

* Add labels to Kubernetes nodes following procedure similar to how it is mentioned in
[osd-topology](../../CRDs/Cluster/ceph-cluster-crd.md#osd-topology).

* (optional) Rook will pass the labels mentioned in [osd-topology](../../CRDs/Cluster/ceph-cluster-crd.md#osd-topology)
as the default set of labels. This can overridden to supply custom labels using the following command.
```console
kubectl patch cm rook-ceph-operator-config -nrook-ceph -p $'data:\n "CSI_CRUSH_LOCATION_LABELS": "<custom-labels>"'
```

Ceph CSI will extract the crush location from the topology labels found on the node
and pass it though krbd options during mapping RBD volumes.

Example:
`"--options read_from_replica=localize,crush_location=zone:east-zone1|region:east"`

> This feature is enabled by default. To disable it,

Reviewer: Do we need to worry about the ceph version or kernel supporting it (back to Pacific) before we enable it by default? Or the topology will be ignored where not supported?

Author: https://docs.ceph.com/en/latest/man/8/rbd/#kernel-rbd-krbd-options

This document tells there's support from  5.8 and
This mode is safe for general use only since Octopus (i.e. after “ceph osd require-osd-release octopus”). Otherwise it should be limited to read-only workloads such as images mapped read-only everywhere or snapshots.

```

* PVCs created using the new storageclass will be encrypted.

> ### Enable Read affinity for RBD volumes

Reviewer: Seems like it should be a higher level title
```suggestion
## Enable Read affinity for RBD volumes
```

> Ceph CSI supports mapping RBD volumes with krbd options

Reviewer: Can you separate out the description from the details? Something like this:
```
Ceph CSI supports mapping RBD volumes with krbd options to
allow serving reads from an OSD in proximity to the client, according to OSD locations
defined in the CRUSH map.

The format of the krbd option is:
`"read_from_replica=localize,crush_location=type1:value1|type2:value2"`

Refer to the [krbd-options](https://docs.ceph.com/en/latest/man/8/rbd/#kernel-rbd-krbd-options)
for more details.
```

`"read_from_replica=localize,crush_location=type1:value1|type2:value2"` to
allow serving reads from the most local OSD (according to OSD locations as
defined in the CRUSH map).
Refer [krbd-options](https://docs.ceph.com/en/latest/man/8/rbd/#kernel-rbd-krbd-options)
for more details.

Execute the following steps:

> * Add labels to Kubernetes nodes following procedure similar to how it is mentioned in

Reviewer: ```suggestion
* Add topology labels to the Kubernetes nodes. The same labels may be used as mentioned in the [OSD topology](../../CRDs/Cluster/ceph-cluster-crd.md#osd-topology) topic.
```

Author: missed this, modified now.

[osd-topology](../../CRDs/Cluster/ceph-cluster-crd.md#osd-topology).

> * Update the `rook-ceph-operator-config` configmap and patch it using the following

Reviewer: ```suggestion
* Patch the `rook-ceph-operator-config` configmap to enable read affinity:
```

command.
```console

> kubectl patch cm rook-ceph-operator-config -nrook-ceph -p $'data:\n "CSI_ENABLE_READ_AFFINITY": "true"'

Reviewer: How about a patch command that just replaces the default `false` value? Otherwise this command seems like it would add a duplicate setting if that setting is already in the configmap.

Author: > How about a patch command that just replaces the default `false` value? Otherwise this command seems like it would add a duplicate setting if that setting is already in the configmap.

that command would fail in case the existing configmap does not have the default being set to false. 
But since we'll have it turned on by default, I'll move this command to the bottom and 
give them instructions to turn it off if required (patch in this case too).

```

* (optional) Rook will pass the labels mentioned in [osd-topology](../../CRDs/Cluster/ceph-cluster-crd.md#osd-topology)
as the default set of labels. This can overridden to supply custom labels using the following command.
```console

> kubectl patch cm rook-ceph-operator-config -nrook-ceph -p $'data:\n "CSI_CRUSH_LOCATION_LABELS": "<custom-labels>"'

Reviewer: f different labels are used from the OSD topology labels, wouldn't the labels be invalid? Or what is an example of the custom labels?

Author: > f different labels are used from the OSD topology labels, wouldn't the labels be invalid? Or what is an example of the custom labels?

In case of an external cluster ?

```

> Ceph CSI will extract the crush location and pass it as

Reviewer: ```suggestion
Ceph CSI will extract the crush location from the topology labels found on the node pass it to the krbd options.

For example: 
`"--options read_from_replica=localize,crush_location=zone:east-zone1|region:east"`
```

`"--options read_from_replica=localize,crush_location=zone:east-zone1|region:east"`
krbd option while mapping RBD volumes.
If enabled, this option will be added to all RBD volumes mapped by Ceph CSI.

!!! note

>     Make sure the version of ceph-csi used is `v3.8.0+`.

Reviewer: The 3.8 driver will come with Rook 1.11 where this feature would be enabled. I don't think we need to mention this version here.
```suggestion
```

>     Label values will have all its dots `"."` normalized with dashes `"-"`

Reviewer: This seems like an implementation detail. Do we need to tell the user? Seems like we can keep it simple and not tell the user.
```suggestion
```
